# Iowa State Prison Recidivism

**source**:
<br><br>
https://data.iowa.gov/Public-Safety/3-Year-Recidivism-for-Offenders-Released-from-Pris/mw8r-vqy4/data
<br><br>
- Statistics about inmate recidivism within 3 years of being released from prison in Iowa.
Covers the time period from 
<br><br>
***
**Features**:
<br><br>
##### Fiscal Year Released:
- Fiscal year (year ending June 30) for which the offender was released from prison.
<br><br>

##### Recidivism Reporting Year:
- Fiscal year (year ending June 30) that marks the end of the 3-year tracking period. For example, offenders exited prison in FY 2012 are found in recidivism reporting year FY 2015.
<br><br>

##### Race-Ethnicity: 
- Offender's Race and Ethnicity
<br><br>

##### Age At Release:
- Offender's age group at release from prison.
<br><br>

##### Convicting Offense Classification: 
- Maximum penalties:
    - *A Felony* : Life 
    - *B Felony* : 25 or 50 years 
    - *C Felony* : 10 years 
    - *D Felony* : 5 years
    - *Aggravated Misdemeanor* : 2 years
    - *Serious Misdemeanor* : 1 year
    - *Simple Misdemeanor* : 30 days
<br><br>

##### Convicting Offense Type: 
- General category for the most serious offense for which the offender was placed in prison.
<br><br>

##### Convicting Offense Subtype: 
-Further classification of the most serious offense for which the offender was placed in prison.
<br><br>

##### Release Type: 
- Reasoning for Offender's release from prison.
<br><br>

##### Main Supervising District:
- The Judicial District supervising the offender for the longest time during the tracking period.
<br><br>

##### Recidivism - Return to Prison:	
- *No* : No Recidivism
- *Yes* : Prison admission for any reason within the 3-year tracking period
<br><br>

##### Days to Recidivism:
- Number of days it took before the offender returned to prison.
<br><br>

##### New Conviction Offense Classification: 
- New conviction maximum penalties: 
    - *A Felony* : Life
    - *B Felony* : 25 or 50 years
    - *C Felony* : 10 years
    - *D Felony* : 5 years
    - *Aggravated Misdemeanor* : 2 years
    - *Serious Misdemeanor* : 1 year
    - *Simple Misdemeanor* : 30 days
<br><br>

##### New Conviction Offense Type: 
- General category for the new conviction while the offender is out of prison.
<br><br>

##### New Conviction Offense Sub Type: 
- Further classification of the new conviction.
<br><br>

##### Part of Target Population: 
- The Department of Corrections has undertaken specific strategies to reduce recidivism rates for prisoners who are on parole.
<br><br>

##### Recidivism Type:
- There is no description of this feature. It pertains only to recidivists and to the category of recidivsm they follow into.
<br><br>

##### Sex:
- Gender of our offender

## Obtain Data
<br><br>
***

In [1]:
# Import all necesary modules using bs_ds
from bs_ds.imports import *
# Set plot style
from jupyterthemes import jtplot
jtplot.style(theme='monokai', context='notebook')

View our documentation at https://bs-ds.readthedocs.io/en/latest/bs_ds.html
For convenient loading of standard modules :
>> from bs_ds.imports import *



Module/Package Handle
pandas                               pd
numpy                                np
matplotlib                          mpl
matplotlib.pyplot                   plt
seaborn                             sns

In [2]:
recidivists = pd.read_csv('3-Year_Recidivism_for_Offenders_Released_from_Prison_in_Iowa.csv')
print(recidivists.shape)
display(recidivists.head())
recidivists.info()

(26020, 17)


Fiscal Year Released  Recidivism Reporting Year      Race - Ethnicity  \
0                  2010                       2013  Black - Non-Hispanic   
1                  2010                       2013  White - Non-Hispanic   
2                  2010                       2013  White - Non-Hispanic   
3                  2010                       2013  White - Non-Hispanic   
4                  2010                       2013  Black - Non-Hispanic   

  Age At Release  Convicting Offense Classification Convicting Offense Type  \
0           25-34                          C Felony                 Violent   
1           25-34                          D Felony                Property   
2           35-44                          B Felony                    Drug   
3           25-34                          B Felony                   Other   
4           35-44                          D Felony                 Violent   

  Convicting Offense Subtype                  Release Type  \
0                    Robbery                        Parole   
1                      Theft  Discharged – End of Sentence   
2                Trafficking                        Parole   
3             Other Criminal                        Parole   
4                    Assault  Discharged – End of Sentence   

  Main Supervising District Recidivism - Return to Prison  Days to Recidivism  \
0                       7JD                           Yes               433.0   
1                       NaN                           Yes               453.0   
2                       5JD                           Yes               832.0   
3                       6JD                            No                 NaN   
4                       NaN                           Yes               116.0   

  New Conviction Offense Classification New Conviction Offense Type  \
0                              C Felony                        Drug   
1                                   NaN                         NaN   
2                                   NaN                         NaN   
3                                   NaN                         NaN   
4                                   NaN                         NaN   

  New Conviction Offense Sub Type Part of Target Population Recidivism Type  \
0                     Trafficking                       Yes             New   
1                             NaN                        No            Tech   
2                             NaN                       Yes            Tech   
3                             NaN                       Yes   No Recidivism   
4                             NaN                        No            Tech   

    Sex  
0  Male  
1  Male  
2  Male  
3  Male  
4  Male

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26020 entries, 0 to 26019
Data columns (total 17 columns):
Fiscal Year Released                     26020 non-null int64
Recidivism Reporting Year                26020 non-null int64
Race - Ethnicity                         25990 non-null object
Age At Release                           26017 non-null object
Convicting Offense Classification        26020 non-null object
Convicting Offense Type                  26020 non-null object
Convicting Offense Subtype               26020 non-null object
Release Type                             24258 non-null object
Main Supervising District                16439 non-null object
Recidivism - Return to Prison            26020 non-null object
Days to Recidivism                       8681 non-null float64
New Conviction Offense Classification    6718 non-null object
New Conviction Offense Type              6718 non-null object
New Conviction Offense Sub Type          6699 non-null object
Part of Target

# Preprocessing
<br><br>
***

There are four features here that only relate to recidivists.  For the initial analysis, that data will be subset for post classification analysis.  
<br><br>
__Seperate data and rename columns__:
<br><br>
Dataset 1 : 
_prison_ 
- All rows and columns except dataset 2 accept those that only pertain to recidivists:<br><br>

Dataset 2 : 
_recidivists_
- Subset of rows with the additional columns added

    - Days to Recidivism
    - New Conviction Offense Classification
    - New Conviction Offense Type
    - New Conviction Offense Sub Type

## Seperate Datasets
<br><br>
***
**Rename columns**

In [3]:
# Get column names to map dictionary of new column titles
colnames_orig = recidivists.columns

# New column names
colnames_short = ('yr_released','report_year','race_ethnicity',
                  'age_released','crime_class','crime_type',
                  'crime_subtype','release_type', 'super_dist',
                  'recidivist','days','new_class','new_type',
                  'new_sub','target_pop','recid_type','sex')

# zip old columsn and new columns together
column_legend = dict(zip(colnames_orig,colnames_short))
column_legend

{'Fiscal Year Released': 'yr_released',
 'Recidivism Reporting Year': 'report_year',
 'Race - Ethnicity': 'race_ethnicity',
 'Age At Release ': 'age_released',
 'Convicting Offense Classification': 'crime_class',
 'Convicting Offense Type': 'crime_type',
 'Convicting Offense Subtype': 'crime_subtype',
 'Release Type': 'release_type',
 'Main Supervising District': 'super_dist',
 'Recidivism - Return to Prison': 'recidivist',
 'Days to Recidivism': 'days',
 'New Conviction Offense Classification': 'new_class',
 'New Conviction Offense Type': 'new_type',
 'New Conviction Offense Sub Type': 'new_sub',
 'Part of Target Population': 'target_pop',
 'Recidivism Type': 'recid_type',
 'Sex': 'sex'}

In [4]:
# Map new names to correct columns
recidivists.rename(mapper=column_legend, axis=1, inplace=True)

# Seperate Columns that pertain only to recidivists
prison = recidivists.drop(['days',
                           'new_class',
                           'new_type', 
                           'new_sub',
                           'recid_type'], axis=1)

### Dataset 1
***
- 26020 rows, each a unique criminal!
- 12 columns

In [5]:
# Dataset 1 
print(prison.shape)
display(prison.head())
display(prison.describe())
prison.info()

(26020, 12)


yr_released  report_year        race_ethnicity age_released crime_class  \
0         2010         2013  Black - Non-Hispanic        25-34    C Felony   
1         2010         2013  White - Non-Hispanic        25-34    D Felony   
2         2010         2013  White - Non-Hispanic        35-44    B Felony   
3         2010         2013  White - Non-Hispanic        25-34    B Felony   
4         2010         2013  Black - Non-Hispanic        35-44    D Felony   

  crime_type   crime_subtype                  release_type super_dist  \
0    Violent         Robbery                        Parole        7JD   
1   Property           Theft  Discharged – End of Sentence        NaN   
2       Drug     Trafficking                        Parole        5JD   
3      Other  Other Criminal                        Parole        6JD   
4    Violent         Assault  Discharged – End of Sentence        NaN   

  recidivist target_pop   sex  
0        Yes        Yes  Male  
1        Yes         No  Male  
2        Yes        Yes  Male  
3         No        Yes  Male  
4        Yes         No  Male

yr_released   report_year
count  26020.000000  26020.000000
mean    2012.600769   2015.600769
std        1.661028      1.661028
min     2010.000000   2013.000000
25%     2011.000000   2014.000000
50%     2013.000000   2016.000000
75%     2014.000000   2017.000000
max     2015.000000   2018.000000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26020 entries, 0 to 26019
Data columns (total 12 columns):
yr_released       26020 non-null int64
report_year       26020 non-null int64
race_ethnicity    25990 non-null object
age_released      26017 non-null object
crime_class       26020 non-null object
crime_type        26020 non-null object
crime_subtype     26020 non-null object
release_type      24258 non-null object
super_dist        16439 non-null object
recidivist        26020 non-null object
target_pop        26020 non-null object
sex               26017 non-null object
dtypes: int64(2), object(10)
memory usage: 2.4+ MB


### Dataset 2 
***
- 8681 rows
- 17 columns
- Might drop a column regarding type if they are redudant

In [6]:
# Dataset 2
recidivists = recidivists.loc[recidivists.recidivist == 'Yes']
print(f"Shape (rows,columns): {recidivists.shape}\n")
display(recidivists.info())

Shape (rows,columns): (8681, 17)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8681 entries, 0 to 26019
Data columns (total 17 columns):
yr_released       8681 non-null int64
report_year       8681 non-null int64
race_ethnicity    8681 non-null object
age_released      8681 non-null object
crime_class       8681 non-null object
crime_type        8681 non-null object
crime_subtype     8681 non-null object
release_type      8331 non-null object
super_dist        6633 non-null object
recidivist        8681 non-null object
days              8681 non-null float64
new_class         6718 non-null object
new_type          6718 non-null object
new_sub           6699 non-null object
target_pop        8681 non-null object
recid_type        8681 non-null object
sex               8681 non-null object
dtypes: float64(1), int64(2), object(14)
memory usage: 1.2+ MB


None

## Scrub 
<br><br>
***
- Dataset 1

In [7]:
# import pandas wrapper tools for inspection and data cleaning
from bs_ds.bamboo import check_null, check_unique, check_numeric
prison.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26020 entries, 0 to 26019
Data columns (total 12 columns):
yr_released       26020 non-null int64
report_year       26020 non-null int64
race_ethnicity    25990 non-null object
age_released      26017 non-null object
crime_class       26020 non-null object
crime_type        26020 non-null object
crime_subtype     26020 non-null object
release_type      24258 non-null object
super_dist        16439 non-null object
recidivist        26020 non-null object
target_pop        26020 non-null object
sex               26017 non-null object
dtypes: int64(2), object(10)
memory usage: 2.4+ MB


###  Missing and Unique Values, Data types
<br><br>
***

In [8]:
# Check null values
check_null(prison)

# of Identified Null Values in "Object" columns:



Null values  Total Values  Percent
Column                                            
yr_released               0         26020    0.000
report_year               0         26020    0.000
race_ethnicity           30         26020    0.115
age_released              3         26020    0.012
crime_class               0         26020    0.000
crime_type                0         26020    0.000
crime_subtype             0         26020    0.000
release_type           1762         26020    6.772
super_dist             9581         26020   36.822
recidivist                0         26020    0.000
target_pop                0         26020    0.000
sex                       3         26020    0.012

__Missing Values__

***

In [9]:
print('NaN is age_released:')
display(prison.loc[prison.age_released.isna()])# These and 'sex' NaN's will drop with 'race_ethnicity'

print("NaN is race_ethincity:")
display(prison.loc[prison.race_ethnicity.isna()])# Drop These

# looking at time blocks that are missing by viewing the two numerical columns for these rows 
print('NaN in super_dist column:')
display(prison.loc[prison.super_dist.isna()].describe())

# how many of the same rows are missing
print('NaN in both super_dist and release_type:')
display(prison.loc[(prison.super_dist.isna())&(prison.release_type.isna())].describe())

print('NaN in only release_type')
display(prison.loc[prison.release_type.isna()])

NaN is age_released:


yr_released  report_year race_ethnicity age_released  \
8542          2012         2015          N/A -          NaN   
9657          2012         2015          N/A -          NaN   
10533         2012         2015          N/A -          NaN   

                  crime_class    crime_type  crime_subtype  \
8542   Aggravated Misdemeanor      Property       Burglary   
9657                 D Felony  Public Order            OWI   
10533  Aggravated Misdemeanor      Property  Forgery/Fraud   

                       release_type super_dist recidivist target_pop  sex  
8542   Discharged – End of Sentence        NaN         No         No  NaN  
9657   Discharged – End of Sentence        NaN         No         No  NaN  
10533  Discharged – End of Sentence        NaN         No         No  NaN

NaN is race_ethincity:


yr_released  report_year race_ethnicity  age_released  \
7636          2012         2015            NaN         35-44   
7665          2012         2015            NaN         25-34   
7794          2012         2015            NaN         45-54   
7913          2012         2015            NaN         25-34   
7966          2012         2015            NaN         35-44   
8144          2012         2015            NaN         25-34   
8145          2012         2015            NaN         45-54   
8350          2012         2015            NaN         25-34   
8975          2012         2015            NaN         45-54   
9071          2012         2015            NaN         45-54   
9386          2012         2015            NaN      Under 25   
9491          2012         2015            NaN         35-44   
9590          2012         2015            NaN         35-44   
9718          2012         2015            NaN         45-54   
9815          2012         2015            NaN         25-34   
9849          2012         2015            NaN      Under 25   
9850          2012         2015            NaN      Under 25   
9894          2012         2015            NaN         45-54   
9900          2012         2015            NaN         25-34   
9946          2012         2015            NaN         25-34   
10291         2012         2015            NaN         45-54   
10636         2012         2015            NaN         25-34   
10694         2012         2015            NaN         25-34   
10726         2012         2015            NaN      Under 25   
10813         2012         2015            NaN         25-34   
10957         2012         2015            NaN         35-44   
11185         2012         2015            NaN         35-44   
11704         2012         2015            NaN         25-34   
12174         2012         2015            NaN  55 and Older   
12311         2012         2015            NaN      Under 25   

                                    crime_class    crime_type  \
7636   Felony - Enhancement to Original Penalty          Drug   
7665                                   C Felony      Property   
7794                     Aggravated Misdemeanor       Violent   
7913                                   D Felony      Property   
7966                     Aggravated Misdemeanor       Violent   
8144                                   D Felony          Drug   
8145                                   D Felony      Property   
8350                                   D Felony       Violent   
8975                     Aggravated Misdemeanor      Property   
9071                                   C Felony       Violent   
9386                                   C Felony       Violent   
9491                     Aggravated Misdemeanor  Public Order   
9590                     Aggravated Misdemeanor       Violent   
9718                                   D Felony  Public Order   
9815                                   C Felony       Violent   
9849                     Aggravated Misdemeanor       Violent   
9850                     Aggravated Misdemeanor       Violent   
9894                                   B Felony          Drug   
9900                     Aggravated Misdemeanor  Public Order   
9946                     Aggravated Misdemeanor  Public Order   
10291                                  C Felony          Drug   
10636                                  B Felony       Violent   
10694                                  D Felony  Public Order   
10726                    Aggravated Misdemeanor         Other   
10813                    Aggravated Misdemeanor      Property   
10957                    Aggravated Misdemeanor  Public Order   
11185                                  C Felony          Drug   
11704                                  C Felony          Drug   
12174                                  D Felony       Violent   
12311                                  D Felony          Drug   

      

NaN in super_dist column:


yr_released  report_year
count  9581.000000  9581.000000
mean   2012.345058  2015.345058
std       1.654413     1.654413
min    2010.000000  2013.000000
25%    2011.000000  2014.000000
50%    2012.000000  2015.000000
75%    2014.000000  2017.000000
max    2015.000000  2018.000000

NaN in both super_dist and release_type:


yr_released  report_year
count       1762.0       1762.0
mean        2014.0       2017.0
std            0.0          0.0
min         2014.0       2017.0
25%         2014.0       2017.0
50%         2014.0       2017.0
75%         2014.0       2017.0
max         2014.0       2017.0

NaN in only release_type


yr_released  report_year  \
17063         2014         2017   
17065         2014         2017   
17066         2014         2017   
17067         2014         2017   
17068         2014         2017   
17069         2014         2017   
17070         2014         2017   
17074         2014         2017   
17077         2014         2017   
17078         2014         2017   
17079         2014         2017   
17080         2014         2017   
17083         2014         2017   
17086         2014         2017   
17087         2014         2017   
17088         2014         2017   
17089         2014         2017   
17090         2014         2017   
17094         2014         2017   
17095         2014         2017   
17100         2014         2017   
17102         2014         2017   
17103         2014         2017   
17104         2014         2017   
17105         2014         2017   
17106         2014         2017   
17108         2014         2017   
17109         2014         2017   
17111         2014         2017   
17112         2014         2017   
...            ...          ...   
21577         2014         2017   
21578         2014         2017   
21579         2014         2017   
21580         2014         2017   
21581         2014         2017   
21583         2014         2017   
21585         2014         2017   
21586         2014         2017   
21588         2014         2017   
21590         2014         2017   
21597         2014         2017   
21600         2014         2017   
21601         2014         2017   
21602         2014         2017   
21603         2014         2017   
21604         2014         2017   
21605         2014         2017   
21606         2014         2017   
21615         2014         2017   
21617         2014         2017   
21621         2014         2017   
21622         2014         2017   
21623         2014         2017   
21625         2014         2017   
21630         2014         2017   
21633         2014         2017   
21634         2014         2017   
21637         2014         2017   
21640         2014         2017   
21641         2014         2017   

                                        race_ethnicity  age_released  \
17063                             White - Non-Hispanic         35-44   
17065                             White - Non-Hispanic      Under 25   
17066                             White - Non-Hispanic      Under 25   
17067                             White - Non-Hispanic      Under 25   
17068                             White - Non-Hispanic      Under 25   
17069                             White - Non-Hispanic      Under 25   
17070                             Black - Non-Hispanic         35-44   
17074                                 White - Hispanic  55 and Older   
17077                             White - Non-Hispanic         35-44   
17078                             White - Non-Hispanic         35-44   
17079                             White - Non-Hispanic         35-44   
17080                             White - Non-Hispanic         35-44   
17083                             Black - Non-Hispanic  55 and Older   
17086                             White - Non-Hispanic  55 and Older   
17087                             White - Non-Hispanic  55 and Older   
17088                             White - Non-Hispanic  55 and Older   
17089                             White - Non-Hispanic  55 and Older   
17090                             White - Non-Hispanic  55 and Older   
17094                             Black - Non-Hispanic         45-54   
17095                             Black - Non-Hispanic         45-54   
17100                             Black - Non-Hispanic         25-34   
17102                             White - Non-Hispanic      Under 25   
17103                             Black - Non-Hispanic      Under 25   
17104                             White - Non-Hispanic         25-34   
17105                             Wh

The columns *super_dist* and *release_type* are each missing significant amounts of values. Together missing over 40% of of rows are missing values in those two columns. Because the missing values in *release_type* all come from rows that are adjacent to eachother, if dropped, the data may lose significance due to the temporal element of the dataset. I am assuming, based on the release year and report year that there is some type of order to this. Even if they were released consecutively in the order of the rows, they still come from the same general time period.  Changing practices within the prison over time, could have an effect on the recidivism rates of ex-cons. For now those rows will stay in the set. The rows corresponding with the missing values from other columns will be dropped.

In [14]:
# drop rows with missing values and affirm drop
prison = prison.loc[(~prison.race_ethnicity.isna())& (~prison.sex.isna())]
check_null(prison)
df = prison

# of Identified Null Values in "Object" columns:



Null values  Total Values  Percent
Column                                            
yr_released               0         25987    0.000
report_year               0         25987    0.000
race_ethnicity            0         25987    0.000
age_released              0         25987    0.000
crime_class               0         25987    0.000
crime_type                0         25987    0.000
crime_subtype             0         25987    0.000
release_type           1762         25987    6.780
super_dist             9549         25987   36.745
recidivist                0         25987    0.000
target_pop                0         25987    0.000
sex                       0         25987    0.000

**Unique Values**

***

In [15]:
check_unique(prison)


yr_released Type: int64
Number unique values: 6


yr_released
2012         4707
2013         4687
2014         4586
2015         4374
2011         3917
2010         3716


report_year Type: int64
Number unique values: 6


report_year
2015         4707
2016         4687
2017         4586
2018         4374
2014         3917
2013         3716


race_ethnicity Type: object
Number unique values: 11


race_ethnicity
White - Non-Hispanic                                      17584
Black - Non-Hispanic                                       6109
White - Hispanic                                           1522
American Indian or Alaska Native - Non-Hispanic             502
Asian or Pacific Islander - Non-Hispanic                    192
Black - Hispanic                                             37
American Indian or Alaska Native - Hispanic                  20
White -                                                      12
Asian or Pacific Islander - Hispanic                          5
Black -                                                       2
N/A -                                                         2


age_released Type: object
Number unique values: 5


age_released
25-34                 9543
35-44                 6217
Under 25              4585
45-54                 4340
55 and Older          1302


crime_class Type: object
Number unique values: 15


crime_class
D Felony                                        10478
C Felony                                         6796
Aggravated Misdemeanor                           4916
B Felony                                         1763
Felony - Enhancement to Original Penalty         1532
Felony - Enhanced                                 220
Serious Misdemeanor                               155
Special Sentence 2005                              98
Felony - Mandatory Minimum                         11
Other Felony                                        6
A Felony                                            4
Simple Misdemeanor                                  3
Other Felony (Old Code)                             2
Sexual Predator Community Supervision               2
Other Misdemeanor                                   1


crime_type Type: object
Number unique values: 5


crime_type
Drug                7908
Property            7364
Violent             5805
Public Order        3601
Other               1309


crime_subtype Type: object
Number unique values: 26


crime_subtype
Trafficking                               6486
Assault                                   3182
Burglary                                  2963
Theft                                     2676
OWI                                       1789
Sex                                       1276
Forgery/Fraud                             1208
Other Criminal                            1191
Drug Possession                           1142
Other Violent                              600
Traffic                                    524
Murder/Manslaughter                        392
Weapons                                    371
Alcohol                                    356
Vandalism                                  347
Robbery                                    338
Other Public Order                         311
Other Drug                                 280
Arson                                      161
Sex Offender Registry/Residency            129
Flight/Escape                               83
Kidnap                                      66
Special Sentence Revocation                 63
Prostitution/Pimping                        38
Stolen Property                              9
Animals                                      6


release_type Type: object
Number unique values: 11


release_type
Parole                                       9810
Parole Granted                               5577
Discharged – End of Sentence                 5007
Discharged - Expiration of Sentence          2335
Released to Special Sentence                  401
Special Sentence                              347
Paroled w/Immediate Discharge                 334
Paroled to Detainer - Out of State            137
Paroled to Detainer - INS                     134
Paroled to Detainer - U.S. Marshall            77
Paroled to Detainer - Iowa                     66


super_dist Type: object
Number unique values: 10


super_dist
5JD                       4982
1JD                       2787
2JD                       1988
8JD                       1556
7JD                       1514
3JD                       1188
6JD                       1098
4JD                        666
ISC                        350
Interstate Compact         309


recidivist Type: object
Number unique values: 2


recidivist
No        17306
Yes        8681


target_pop Type: object
Number unique values: 2


target_pop
Yes       14274
No        11713


sex Type: object
Number unique values: 2


sex
Male    22651
Female   3336

**Drop hidden N/A's and other suspiscious rows**
***
- **race_ethnicity** : White-, Black-, N/A-
    All of those labels are incomplete or belong somewhere else.

### Encoding Boolean Values
<br><br>
***

In [ ]:
bools = ['recidivist', 'target_pop', 'sex']

#  EDA
<br><br>
**visualizations**
***

In [ ]:
from bs_ds.bamboo import LabelLibrary, plot_hist_scat, multiplot

# Make list of variables to label, leaving out years for eda
df_code = df
to_encode = [i for i in df_code.columns if i  != 'yr_released' and i != 'report_year']

# fit and transform labels with LabelEncoder wrapper class
labels = LabelLibrary().fit_transform(df_code, to_encode)
print(labels.shape)
labels.head()

In [ ]:
plot_hist_scat(labels, target='recidivist')
plt.show()

In [ ]:
multiplot(labels)
plt.show()


### One Hot Encoded
<br><br>
***

In [ ]:
dummies = pd.get_dummies(df_code, sparse=True)

In [ ]:
dummies.head()